lib/osf-components/addon/components/resources-list/edit-resource/component.ts

```diff
@@ -0,0 +1,104 @@
+import DS from 'ember-data';
+import IntlService from 'ember-intl/services/intl';
+import RegistrationModel from 'ember-osf-web/models/registration';
+import ResourceModel, { ResourceTypes } from 'ember-osf-web/models/resource';
+import buildChangeset from 'ember-osf-web/utils/build-changeset';
+import { tracked } from 'tracked-built-ins';
+
+interface Args {
+    resource?: ResourceModel;
+    registration?: RegistrationModel;
+    reload?: () => {};
+}
+
+interface ResourceValidations {
+    pid: string;
+    description: string;
+    resourceType: string;
+}
+
+export default class EditResourceModal extends Component<Args> {
+    @service toast!: Toastr;
+    @service intl!: IntlService;
+    @service store!: DS.Store;
+
+    availableTypes = Object.values(ResourceTypes);
+    resourceValidations: ValidationObject<ResourceValidations> = {
+        pid: [validatePresence({
+            type: 'invalid',
+            presence: true,
+            translationArgs: { description: this.intl.t('osf-components.resources-list.edit_resource.doi') },
+        })],
+        resourceType: [validatePresence({
+            type: 'blank',
+            presence: true,
+            translationArgs: { description: this.intl.t('osf-components.resources-list.edit_resource.output_type') },
+        })],
+    };
+
+    @tracked changeset: any;
+    @tracked resource?: ResourceModel = this.args.resource;
+    @tracked shouldShowPreview = false;
+
+    @task
+    async onOpen() {
+        if (!this.resource) {
+            this.resource = await this.store.createRecord('resource', { registration: this.args.registration });
+            await this.resource.save();
+        }
+        this.changeset = buildChangeset(this.resource, this.resourceValidations);
+    }
+
+    @action
+    onClose() {
+        if (!this.args.resource) {
+            this.resource = undefined;
+            this.changeset = undefined;
+        }
+        this.shouldShowPreview = false;
+    }
+
+    @task
+    async goToPreview() {
+        this.changeset.validate();
+        if (this.changeset.get('isValid')) {
+            try {
+                await this.changeset.save();
+                this.shouldShowPreview = true;
+            } catch (e) {
+                if (e.errors[0].status === '400') {
+                    this.changeset.addError(
+                        'pid',
+                        this.intl.t(
+                            'validationErrors.invalid',
+                            {
+                                description: this.intl.t('osf-components.resources-list.edit_resource.doi'),
+                            },
+                        ),
+                    );
+                }
+            }
+        }
+    }
+
+    @action
+    goToEdit() {
+        this.shouldShowPreview = false;
+    }
+
+    @task
+    async finalize() {
+        if (this.resource) {
+            this.resource.finalized = true;
+            await this.resource.save();
+        }
+    }
```

Comment:
I think this is supposed to show toast messages in the success and failure cases.